src/ModelContextProtocol.AspNetCore/AuthorizationMiddleware.cs

```diff
@@ -0,0 +1,113 @@
+using Microsoft.AspNetCore.Http;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Options;
+using ModelContextProtocol.Protocol.Auth;
+using ModelContextProtocol.Server;
+using ModelContextProtocol.Utils.Json;
+using System.Text.Json;
+
+namespace ModelContextProtocol.AspNetCore;
+
+/// <summary>
+/// Middleware that handles authorization for MCP servers.
+/// </summary>
+internal class AuthorizationMiddleware
+{
+    private readonly RequestDelegate _next;
+    private readonly ILogger<AuthorizationMiddleware> _logger;
+
+    /// <summary>
+    /// Initializes a new instance of the <see cref="AuthorizationMiddleware"/> class.
+    /// </summary>
+    /// <param name="next">The next middleware in the pipeline.</param>
+    /// <param name="logger">The logger factory.</param>
+    public AuthorizationMiddleware(RequestDelegate next, ILogger<AuthorizationMiddleware> logger)
+    {
+        _next = next ?? throw new ArgumentNullException(nameof(next));
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    }
+
+    /// <summary>
+    /// Processes a request.
+    /// </summary>
+    /// <param name="context">The HTTP context.</param>
+    /// <param name="serverOptions">The MCP server options.</param>
+    /// <param name="authProvider">The authorization provider.</param>
+    /// <returns>A <see cref="Task"/> representing the asynchronous operation.</returns>
+    public async Task InvokeAsync(
+        HttpContext context, 
+        IOptions<McpServerOptions> serverOptions,
+        IServerAuthorizationProvider? authProvider = null)
+    {
+        // Check if authorization is configured
+        if (authProvider == null)
+        {
+            // Authorization is not configured, proceed to the next middleware
+            await _next(context);
+            return;
+        }
+
+        // Handle the PRM document endpoint
+        if (context.Request.Path.StartsWithSegments("/.well-known/oauth-protected-resource"))
+        {
+            _logger.LogDebug("Serving Protected Resource Metadata document");
+            context.Response.ContentType = "application/json";
+            await JsonSerializer.SerializeAsync(
+                context.Response.Body, 
+                authProvider.GetProtectedResourceMetadata(),
+                McpJsonUtilities.DefaultOptions.GetTypeInfo(typeof(ProtectedResourceMetadata)));
+            return;
+        }
+
+        // Serve SSE and message endpoints with authorization
+        if (context.Request.Path.StartsWithSegments("/sse") || 
+            (context.Request.Path.Value?.EndsWith("/message") == true))
```

Comment:
This would probably be better done with [endpoint filters](https://learn.microsoft.com/en-us/aspnet/core/fundamentals/minimal-apis/min-api-filters?view=aspnetcore-9.0) on the RouteGroups we create in MapMcp where the routing has already been done for us. Feel free to ping me if you want any guidance on this.